File: tpl-backend/tpl/routers/logging.py
from fastapi import APIRouter, Depends, HTTPException
from asyncpg import Connection

from tpl.db import get_connection
from tpl.models import (
    StepExecution,
    ExecutionCreate,
    ExecutionComplete,
    ExecutionUpdate,
    IncidentCreate,
    IncidentResolve,
    AdhocCreate,
    ExecutionStats,
)
from tpl.services import logging_service

router = APIRouter(tags=["logging"])


@router.get("/projects/{project_id}/executions", response_model=list[StepExecution])
async def list_executions(project_id: str, db: Connection = Depends(get_connection)):
    return await logging_service.list_executions(db, project_id)


@router.get("/projects/{project_id}/executions/current", response_model=StepExecution | None)
async def get_current_execution(project_id: str, db: Connection = Depends(get_connection)):
    return await logging_service.get_current_execution(db, project_id)


@router.post("/projects/{project_id}/executions/start", response_model=StepExecution, status_code=201)
async def start_execution(project_id: str, body: ExecutionCreate, db: Connection = Depends(get_connection)):
    if not body.plan_step_id:
        raise HTTPException(status_code=400, detail="plan_step_id is required")
    return await logging_service.start_execution(
        db, project_id, body.plan_step_id, body.execution_number,
    )


@router.post("/projects/{project_id}/executions/{execution_id}/complete", response_model=StepExecution)
async def complete_execution(
    project_id: str,
    execution_id: str,
    body: ExecutionComplete,
    db: Connection = Depends(get_connection),
):
    return await logging_service.complete_execution(
        db, project_id, execution_id, body.completion_check, body.notes,
    )


@router.post("/projects/{project_id}/executions/{execution_id}/skip", response_model=StepExecution)
async def skip_execution(
    project_id: str,
    execution_id: str,
    body: ExecutionComplete,
    db: Connection = Depends(get_connection),
):
    return await logging_service.skip_execution(db, project_id, execution_id, body.notes)


@router.post("/projects/{project_id}/executions/incident", response_model=StepExecution, status_code=201)
async def create_incident(project_id: str, body: IncidentCreate, db: Connection = Depends(get_connection)):
    return await logging_service.create_incident(db, project_id, body.plan_step_id, body.notes)


@router.post("/projects/{project_id}/executions/incident/{incident_id}/resolve", response_model=StepExecution)
async def resolve_incident(
    project_id: str,
    incident_id: str,
    body: IncidentResolve,
    db: Connection = Depends(get_connection),
):
    return await logging_service.resolve_incident(
        db, project_id, incident_id, body.reason, body.category, body.notes,
    )


@router.post("/projects/{project_id}/executions/adhoc", response_model=StepExecution, status_code=201)
async def create_adhoc(project_id: str, body: AdhocCreate, db: Connection = Depends(get_connection)):
    return await logging_service.create_adhoc(db, project_id, body.title, body.description, body.notes)


@router.get("/projects/{project_id}/executions/stats", response_model=ExecutionStats)
async def get_execution_stats(project_id: str, db: Connection = Depends(get_connection)):
    return await logging_service.get_execution_stats(db, project_id)

File: tpl-frontend/src/stores/logging.ts
import { writable } from "svelte/store";
import type { StepExecution, ExecutionStats } from "../types";
import { loggingApi } from "../lib/api";

function createLoggingStore() {
  const { subscribe, set, update } = writable<{
    currentProjectId: string | null;
    currentExecution: StepExecution | null;
    executions: StepExecution[];
    incidentActive: boolean;
    stats: ExecutionStats | null;
    loading: boolean;
    error: string | null;
  }>({
    currentProjectId: null,
    currentExecution: null,
    executions: [],
    incidentActive: false,
    stats: null,
    loading: false,
    error: null,
  });

  async function loadProject(projectId: string) {
    update((s) => ({ ...s, currentProjectId: projectId, loading: true, error: null }));
    try {
      const [executions, current] = await Promise.all([
        loggingApi.list(projectId),
        loggingApi.current(projectId).catch(() => null),
      ]);
      update((s) => ({
        ...s,
        executions,
        currentExecution: current,
        incidentActive: current?.type === "incident" && current.status === "in_progress",
        loading: false,
      }));
    } catch (e) {
      update((s) => ({ ...s, loading: false, error: String(e) }));
    }
  }

  async function loadStats(projectId: string) {
    try {
      const stats = await loggingApi.stats(projectId);
      update((s) => ({ ...s, stats }));
    } catch (e) {
      console.error("Failed to load stats:", e);
    }
  }

  async function startStep(projectId: string, planStepId: string, executionNumber = 1): Promise<StepExecution | null> {
    try {
      const execution = await loggingApi.start(projectId, { plan_step_id: planStepId, execution_number: executionNumber });
      await loadProject(projectId);
      return execution;
    } catch (e) {
      update((s) => ({ ...s, error: String(e) }));
      return null;
    }
  }

  async function completeStep(projectId: string, executionId: string, completionCheck?: string, notes?: string) {
    try {
      await loggingApi.complete(projectId, executionId, { completion_check: completionCheck, notes });
      await loadProject(projectId);
    } catch (e) {
      update((s) => ({ ...s, error: String(e) }));
    }
  }

  async function skipStep(projectId: string, executionId: string, notes?: string) {
    try {
      await loggingApi.skip(projectId, executionId, { notes });
      await loadProject(projectId);
    } catch (e) {
      update((s) => ({ ...s, error: String(e) }));
    }
  }

  async function createIncident(projectId: string, planStepId?: string, notes?: string) {
    try {
      await loggingApi.incident.create(projectId, { plan_step_id: planStepId, notes });
      await loadProject(projectId);
    } catch (e) {
      update((s) => ({ ...s, error: String(e) }));
    }
  }

  async function resolveIncident(projectId: string, incidentId: string, reason: string, category?: string, notes?: string) {
    try {
      await loggingApi.incident.resolve(projectId, incidentId, { reason, category, notes });
      await loadProject(projectId);
    } catch (e) {
      update((s) => ({ ...s, error: String(e) }));
    }
  }

  async function createAdhoc(projectId: string, title: string, description?: string, notes?: string) {
    try {
      await loggingApi.adhoc(projectId, { title, description, notes });
      await loadProject(projectId);
    } catch (e) {
      update((s) => ({ ...s, error: String(e) }));
    }
  }

  return {
    subscribe,
    loadProject,
    loadStats,
    startStep,
    completeStep,
    skipStep,
    createIncident,
    resolveIncident,
    createAdhoc,
  };
}

export const loggingStore = createLoggingStore();
